### PR TITLE
Symmetric

### PR DESCRIPTION
**Pull request by cristovaoth**

### CODE REVIEW

File: src/balancerV1/factory.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    const tokens = currentTokensCall.value;
+    if (tokens.some(token => token.equals(GNO_ADDRESS))) {
+      const pool = new WeightedPool(address.toHexString());
+      pool.positions = [];
+      pool.totalSupply = ZERO_BI;
+      pool.gnoBalance = ZERO_BI;
+      pool.gnoIsFirst = false;
+      pool.save();
+      log.info("instantiated WeightedPool instance: {}", [
+        // pool id ?
```
Comment (jfschwarz) [inline src/balancerV1/factory.ts]: yes

File: src/helpers.ts [hunk 2]
```diff
@@ -56,3 +57,222 @@ export function removeOrSaveUser(user: User): void {
...
+      [
+        id,
+        event.transaction.hash.toHexString(),
+        event.transactionLogIndex.toString(),
+      ]
+    );
+    return;
+  }
+
+  // Swap() is emitted after Sync(), so the balance should be up to date
```
Comment (jfschwarz) [inline src/helpers.ts:82]: How does the situation look for balancer? Is Swap emitted after pool.gnoBalance has been updated already?

Comment (jfschwarz) [reply src/helpers.ts:82]: It seems like we don't update `gnoBalance` for Balancer V1 anywhere else, so this won't quite work. It's rather:

```
const gnoReservesBefore = pool.gnoBalance;
const gnoReserves = gnoReserves.plus(gnoIn).minus(gnoOut);
```

**Review by jfschwarz — CHANGES_REQUESTED** [review]

File: src/helpers.ts [hunk 3]
```diff
@@ -56,3 +57,222 @@ export function removeOrSaveUser(user: User): void {
...
+}
+
+export function weightedPoolTransfer(
+  event: ethereum.Event,
+  id: string,
+  from: Address,
+  to: Address,
+  value: BigInt
+): void {
+  //const id = event.address.toHexString();
```
Comment (jfschwarz) [inline src/helpers.ts]: that would be nice, to get rid of `id` param

Comment (jfschwarz) [reply src/helpers.ts]: actually it would be enough to pass the event, no? should be generally, ERC20 Transfer

File: src/helpers.ts (hunk 2 above)
Comment (jfschwarz) [reply src/helpers.ts:82]: ok, now let's adjust the comment, so we remember this.

File: src/helpers.ts (hunk 3 above)
Comment (cristovaoth) [reply src/helpers.ts]: For balancerV1, they don't use the ERC20 event. Its an event with different properties

File: src/helpers.ts (hunk 2 above)
Comment (jfschwarz) [reply src/helpers.ts:82]:
```suggestion
  // Swap() is emitted after Sync(), so the balance should be up to date
  // For Balancer V1 we apply the delta to gnoBalance upfront, so we can reuse this function
```

Comment (jfschwarz) [reply src/helpers.ts:82]: shouldn't we also be able to use this for Balancer V2 now?

File: src/helpers.ts (hunk 3 above)
Comment (jfschwarz) [reply src/helpers.ts:149]: let's get it to this signature:

```
weightedPoolTransfer(event: TransferEvent)
```

File: subgraph-gnosis-chain.yaml [hunk 4]
```diff
@@ -235,3 +257,49 @@ templates:
...
+        #   handler: handleFinalize
+        # - event: LOG_CALL(indexed bytes4,indexed address,bytes)
+        #   topic0: "0x3fdddaa200000000000000000000000000000000000000000000000000000000"
+        #   handler: handleRebind
+        # - event: LOG_CALL(indexed bytes4,indexed address,bytes)
+        #   topic0: "0xe4e1e53800000000000000000000000000000000000000000000000000000000"
+        #   handler: handleRebind
+        # - event: LOG_CALL(indexed bytes4,indexed address,bytes)
+        #   topic0: "0xcf5e7bd300000000000000000000000000000000000000000000000000000000"
+        #   handler: handleUnbind
```
Comment (jfschwarz) [inline subgraph-gnosis-chain.yaml]:
```suggestion
```

cleanup

File: src/helpers.ts (hunk 2 above)
Comment (cristovaoth) [reply src/helpers.ts:99]: I think so. Let me push some changes, and then i adapt it for balancerv2 to use the helper as well

File: subgraph-gnosis-chain.yaml [hunk 5]
```diff
@@ -208,6 +208,28 @@ dataSources:
...
         - event: PairCreated(indexed address,indexed address,address,uint256)
           handler: handleNewPair
+  - kind: ethereum/contract
+    name: ds-balancer-v1-factory
+    network: xdai
+    source:
+      # symmetric's factory address, a balancer-v1 clone
+      address: "0x9B4214FD41cD24347A25122AC7bb6B479BED72Ac"
+      abi: Factory
+      startBlock: 13108160
```
Comment (jfschwarz) [inline subgraph-gnosis-chain.yaml]:
```suggestion
      startBlock: 16465037
```

https://blockscout.com/xdai/mainnet/tx/0x20020bccdaf2361e40834f6530b10fbb142ddfe1339519329dc9e5ba7a3a35e3

File: src/helpers.ts (hunk 2 above)
Comment (cristovaoth) [reply src/helpers.ts:82]: The comment is a good idea.

I think yes we can also use the helper for balanceV2. Will push

File: subgraph-gnosis-chain.yaml (hunk 4 above)
Comment (cristovaoth) [reply subgraph-gnosis-chain.yaml:297]: You checked that none of those events are necessary?

**Pull request merged**